simpleline/render/widget_containers.py

```diff
@@ -0,0 +1,349 @@
+# GNU General Public License along with this program; if not, write to the
+# Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+# 02110-1301, USA.  Any Red Hat trademarks that are incorporated in the
+# source code or documentation are not subject to the GNU General Public
+# License and may only be used or replicated with the express permission of
+# Red Hat, Inc.
+#
+
+from math import ceil
+
+from simpleline.render.widgets import Widget, TextWidget
+
+__all__ = ["ListRowWidget", "ListColumnWidget", "WidgetContainer", "KeyPattern"]
+
+
+class ListRowWidget(Widget):
+    """Place widgets in rows automatically.
+
+    Compared to the ColumnWidget this is able to handle word wrapping correctly.
+
+    Widgets will be placed based on the number of columns in the following way:
+
+    w1   w2   w3
+    w4   w5   w6
+    ....
+    """
+
+    def __init__(self, columns, widgets=None, columns_width=25, spacing=3):
+        """Create ListWidget with specific number of columns.
+
+        :param columns: How many columns we want.
+        :type columns: int, bigger than 0
+
+        :param widgets: List of `WidgetContainer`s. This will be positioned in the ListWidget.
+        :type widgets: Array of `Widget` based class.
+
+        :param columns_width: Width of every column.
+        :type columns_width: int
+
+        :param spacing: Set the spacing between columns.
+        :type spacing: int
+        """
+        super().__init__()
+        self._columns = columns
+
+        if widgets is None:
+            widgets = []
+        self._widgets = widgets
+
+        self._columns_width = columns_width
+        self._spacing = spacing
+        self._key_pattern = None
+        self._numbering_widgets = []
+
+    @property
+    def size(self):
+        """Return items count."""
+        return len(self._widgets)
+
+    @property
+    def key_pattern(self):
+        """Return key pattern which will be used for items numbering.
+
+        Will return `None` if not set.
+        """
+        return self._key_pattern
+
+    @key_pattern.setter
+    def key_pattern(self, key_pattern):
+        """Set the key pattern object which will be used for items numbering.
+
+        Setting `None` will stop doing numbering.
+        """
+        self._key_pattern = key_pattern
+
+    def add_container(self, widget_container):
+        """Add new widget container.
+
+        :param widget_container: Container for widget with callback and data.
+        :type widget_container: The `simpleline.render.widgets.WidgetContainer` class."""
+        self._widgets.append(widget_container)
+
+    def get_container(self, container_id):
+        """Get a widget container.
+
+        :param container_id: ID of the container in the list.
+        :type container_id: int
+
+        :returns: WidgetContainer class.
+        :raise: IndexError
+        """
+        return self._widgets[container_id]
+
+    def modify_prompt(self, prompt):
```

Comment:
I thought it could be valuable. However if there is no demand for this feature, it could be removed now to simplify code base.